README.md
Easy Multilingual website
=================


http://easymultilingual.com


changelog.txt
== Changelog ==

This file contains only old changelog. See readme.txt for newer versions.

= 1.4.5 (2014-04-19) =

* Fix: Notice when combined with WPSEO 1.5+
* Fix: Impossible to disable a widget language filter once set (introduced in 1.4.4)
* Fix: Unexpected redirection of the homepage with language code when permalink structure has no trailing slash (introduced in 1.4.4)
* Fix: Some installs lead to wrong redirection when using domains (introduced in 1.4.4)
* Fix: Possible infinite redirection while previewing posts (introduced in 1.4.4)
* Fix: Uploaded medias don't get a language since WP 3.9
* Fix: Compatibility with Twenty Fourteen Ephemera widget in the version shipped with WP 3.9

= 1.4.4 (2014-04-09) =

* Add: Compatibility with widgets customizer introduced in WP 3.9
* Fix: No post in translation dropdown after switching the language in edit post (introduced in 1.4.3)
* Fix: No canonical redirection when there is no language code in url and the language code is not hidden for the default language
* Fix: Suppress language cookie when using multiple domains

= 1.4.3 (2014-03-22) =

* Add: Serbian translation contributed by Sinisa
* Add: Myanmar translation contributed by Sithu Thwin
* Fix: comment form redirects to wp-admin when using multiple domains or subdomains.
* Fix: fatal error with old versions of PHP (tested on PHP 5.2.4)
* Fix: Bad gateway experienced by users hosted by wpengine.com
* Fix: links got from tiny MCE link button are filtered with admin language filter instead of current post language
* Fix: possibly wrong redirection in check_language_code_in_url when using multiple domains or subdomains

= 1.4.2 (2014-02-24) =

* Add: check multiple post types in EML_Model::count_posts
* Fix: error 404 on category links when setting the language by content (introduced in 1.4.1)
* Fix: PHP notices in frontend-nav-menu.php with Artisteer themes
* Fix: decrease the memory usage of untranslated posts list
* Fix: home page not correctly redirected to canonical when using page on front and page name is kept in url

= 1.4.1 (2014-02-16) =

* Add: Czech translation contributed by [Přemysl Karbula](http://www.premyslkarbula.cz)
* Fix: the displayed language is not correct in quick edit for categories and post tags
* Fix: the language switcher does not display the correct link for translated parent categories if only children have posts
* Fix: 3rd parameter of icl_object_id is not optional
* Fix: issue when combining multiple domains and browser detection -> the combination is now forbidden
* Fix: conflict Shiba Media Library: link between media translations is lost when using media quick edit
* Fix: notice when using taxonomies in wpml-config.xml
* Fix: incorrect post format link
* Fix: Twenty Fourteen Ephemera widget strings are not translated

= 1.4 (2014-01-22) =

* Add Traditionial Chinese translation contributed by [香腸](http://sofree.cc/)
* Minimum WordPress version is now v3.5
* Refresh translations metaboxes: now translated posts are chosen in a dropdown list
* Check if translated archives for category, tag and post format are empty before displaying the language switcher
* Add specific management of translated featured tag in Twenty Fourteen
* Add the possibility not to cache homepage urls with option EML_CACHE_HOME_URL (for users having several domains).
* The function get_pages is now filtered by language
* Ajax requests on frontend are now automatically detected. It is no more necessary to set 'eml_load_front' :)
* Various performance improvements
* 'eml_get_post_types' and 'eml_get_taxonomies' filters must be added *before* 'after_setup_theme' is fired
* Pre 1.2 data will be removed from DB at first upgrade at least 60 days after upgrade to 1.4
* Removed some duplicate code between admin and frontend
* Bug correction: incorrect pagination when using domains or subdomains
* Bug correction: post format link not translated
* Bug correction: impossible to use child terms with same name in hierarchical taxonomies
* Bug correction: the terms list table is filtered according to new translation language instead of admin language filter

= 1.3.1 (2013-12-13) =

* Bug correction: fatal error on settings page if a static front page without language is set
* Bug correction: wrong home url when using different domains per language

= 1.3 (2013-12-11) =

* Refresh admin UI for better look in WP 3.8 and more dynamic comportment
* The "Detect browser language" option does now also controls returning visits (based on cookie).
* Improved performance by optimizing some queries for WP 3.5+
* The user biography in default language is now stored in default WordPress usermeta
* Add language parameter in API function eml_home_url and allow to call it on admin side
* Calling 'get_terms' with the 'lang' parameter now uses a cache object per language
* Bug correction: conflict with unstranslated taxonomies
* Bug correction: possible malformed translation archive url in language switcher
* Bug correction: a wrong language may be displayed in quick edit dropdown
* Bug correction: it is possible to add multiple translations (in the same language) for a single taxonomy term
* Bug correction: non public post types and taxonomies are visible in EasyMultilingual settings
* Bug correction: the language is always chosen from cookie (or browser preferences) in some installations
* Bug correction: Firefox language preference is not recognized when comparison is made on locale (instead of ISO 639-1 language code)
* Bug correction: incorrect tax_query in EML_Auto_Translate

= 1.2.4 (2013-11-28) =

* Better support for theme customizer
* Bug correction: admin bar search does not filter by language
* Bug correction: possible conflict on secondary query when querying taxonomies or single page
* Bug correction: post type is not included in url when editing or adding a term translation
* Bug correction: various warnings and PHP notices

= 1.2.3 (2013-11-17) =


admin/settings.php
<?php

/*
 * a class for the EasyMultilingual settings pages
 *
 * @since 1.2
 */
class EML_Settings {
	public $links_model, $model, $options;
	protected $active_tab, $strings = array(); // strings to translate

	/*
	 * constructor
	 *
	 * @since 1.2
	 *
	 * @param object $easyMultilingual
	 */
	public function __construct(&$easyMultilingual) {
		$this->links_model = &$easyMultilingual->links_model;
		$this->model = &$easyMultilingual->model;
		$this->options = &$easyMultilingual->options;

		$this->active_tab = !empty($_GET['tab']) ? $_GET['tab'] : 'lang';

		// adds screen options and the about box in the languages admin panel
		add_action('load-settings_page_mlang',  array(&$this, 'load_page'));

		// saves per-page value in screen option
		add_filter('set-screen-option', create_function('$s, $o, $v', 'return $v;'), 10, 3);
	}

	/*
	 * adds screen options and the about box in the languages admin panel
	 *
	 * @since 0.9.5
	 */
	public function load_page() {
		// test of $this->active_tab avoids displaying the automatically generated screen options on other tabs
		switch ($this->active_tab) {
			case 'lang':
				if (!defined('EML_DISPLAY_ABOUT') || EML_DISPLAY_ABOUT) {
					add_meta_box(
						'eml_about_box',
						__('About EasyMultilingual', 'easyMultilingual'),
						create_function('', "include(EML_ADMIN_INC.'/view-about.php');"),
						'settings_page_mlang',
						'normal'
					);
				}

				add_screen_option('per_page', array(
					'label'   => __('Languages', 'easyMultilingual'),
					'default' => 10,
					'option'  => 'eml_lang_per_page'
				));
				break;

			case 'strings':
				add_screen_option('per_page', array(
					'label'   => __('Strings translations', 'easyMultilingual'),
					'default' => 10,
					'option'  => 'eml_strings_per_page'
				));
				break;

			default:
				break;
		}
	}

	/*
	 * diplays the 3 tabs pages: languages, strings translations, settings
	 * also manages user input for these pages
	 *
	 * @since 0.1
	 */
	public function languages_page() {
		// prepare the list of tabs
		$tabs = array('lang' => __('Languages','easyMultilingual'));

		// only if at least one language has been created
		if ($listlanguages = $this->model->get_languages_list()) {
			$tabs['strings'] = __('Strings translation','easyMultilingual');
			$tabs['settings'] = __('Settings', 'easyMultilingual');
		}

		$tabs = apply_filters('eml_settings_tabs', $tabs);

		switch($this->active_tab) {
			case 'lang':
				// prepare the list table of languages
				$list_table = new EML_Table_Languages();
				$list_table->prepare_items($listlanguages);
				break;

			case 'strings':
				// get the strings to translate
				$data = $this->get_strings();

				$selected = empty($_REQUEST['group']) ? -1 : $_REQUEST['group'];
				foreach ($data as $key=>$row) {
					$groups[] = $row['context']; // get the groups

					// filter for search string
					if (($selected !=-1 && $row['context'] != $selected) || (!empty($_REQUEST['s']) && stripos($row['name'], $_REQUEST['s']) === false && stripos($row['string'], $_REQUEST['s']) === false))
						unset ($data[$key]);
				}

				$groups = array_unique($groups);

				// load translations
				foreach ($listlanguages as $language) {
					// filters by language if requested
					if (($lg = get_user_meta(get_current_user_id(), 'eml_filter_content', true)) && $language->slug != $lg)
						continue;

					$mo = new EML_MO();
					$mo->import_from_db($language);
					foreach ($data as $key=>$row) {
						$data[$key]['translations'][$language->name] = $mo->translate($row['string']);
						$data[$key]['row'] = $key; // store the row number for convenience
					}
				}

				$string_table = new EML_Table_String($groups, $selected);
				$string_table->prepare_items($data);
				break;

			case 'settings':
				$post_types = get_post_types(array('public' => true, '_builtin' => false));
				$post_types = array_diff($post_types, get_post_types(array('_pll' => true)));
				$post_types = array_unique(apply_filters('eml_get_post_types', $post_types, true));

				$taxonomies = get_taxonomies(array('public' => true, '_builtin' => false));
				$taxonomies = array_diff($taxonomies, get_taxonomies(array('_pll' => true)));
				$taxonomies = array_unique(apply_filters('eml_get_taxonomies', $taxonomies , true));
				break;

			default:
				break;
		}

		$using_permalinks = $GLOBALS['wp_rewrite']->using_permalinks();

		$action = isset($_REQUEST['eml_action']) ? $_REQUEST['eml_action'] : '';

		switch ($action) {
			case 'add':
				check_admin_referer( 'add-lang', '_wpnonce_add-lang' );

				if ($this->model->add_language($_POST))
					EML_Admin::download_mo($_POST['locale']);

				$this->redirect(); // to refresh the page (possible thanks to the $_GET['noheader']=true)
				break;

			case 'delete':
				check_admin_referer('delete-lang');

				if (!empty($_GET['lang']))
					$this->model->delete_language((int) $_GET['lang']);

				$this->redirect(); // to refresh the page (possible thanks to the $_GET['noheader']=true)
				break;

			case 'edit':
				if (!empty($_GET['lang']))
					$edit_lang = $this->model->get_language((int) $_GET['lang']);
				break;

			case 'update':
				check_admin_referer( 'add-lang', '_wpnonce_add-lang' );

				$error = $this->model->update_language($_POST);

				$this->redirect(); // to refresh the page (possible thanks to the $_GET['noheader']=true)
				break;

			case 'string-translation':
				if (!empty($_REQUEST['submit'])) {
					check_admin_referer( 'string-translation', '_wpnonce_string-translation' );
					$strings = $this->get_strings();

					foreach ($this->model->get_languages_list() as $language) {
						if(empty($_POST['translation'][$language->name])) // in case the language filter is active (thanks to John P. Bloch)
							continue;

						$mo = new EML_MO();
						$mo->import_from_db($language);

						foreach ($_POST['translation'][$language->name] as $key=>$translation)
							$mo->add_entry($mo->make_entry($strings[$key]['string'], stripslashes($translation)));

						// clean database (removes all strings which were registered some day but are no more)
						if (!empty($_POST['clean'])) {
							$new_mo = new EML_MO();

							foreach ($strings as $string)
								$new_mo->add_entry($mo->make_entry($string['string'], $mo->translate($string['string'])));
						}

						isset($new_mo) ? $new_mo->export_to_db($language) : $mo->export_to_db($language);
					}
					add_settings_error('general', 'eml_strings_translations_updated', __('Translations updated.', 'easyMultilingual'), 'updated');
				}

				do_action('eml_save_strings_translations');

				// unregisters strings registered through WPML API
				if ($string_table->current_action() == 'delete' && !empty($_REQUEST['strings']) && function_exists('icl_unregister_string')) {
					check_admin_referer( 'string-translation', '_wpnonce_string-translation' );
					$strings = $this->get_strings();

					foreach ($_REQUEST['strings'] as $key)
						icl_unregister_string($strings[$key]['context'], $strings[$key]['name']);
				}

				// to refresh the page (possible thanks to the $_GET['noheader']=true)
				$this->redirect(array_intersect_key($_REQUEST, array_flip(array('s', 'paged', 'group'))));
				break;

			case 'options':
				check_admin_referer( 'options-lang', '_wpnonce_options-lang' );

				$this->options['default_lang'] = sanitize_title($_POST['default_lang']); // we have slug as value

				foreach(array('force_lang', 'rewrite') as $key)
					$this->options[$key] = isset($_POST[$key]) ? (int) $_POST[$key] : 0;

				// FIXME : TODO error message if not a valid url
				if (3 == $this->options['force_lang'] && isset($_POST['domains']) && is_array($_POST['domains'])) {
					foreach ($_POST['domains'] as $key => $domain) {
						$this->options['domains'][$key] = esc_url_raw(trim($domain));
					}
					$this->options['domains'][$this->options['default_lang']] = get_option('home');
				}

				foreach (array('browser', 'hide_default', 'redirect_lang', 'media_support') as $key)
					$this->options[$key] = isset($_POST[$key]) ? 1 : 0;

				if (3 == $this->options['force_lang'])
					$this->options['browser'] = $this->options['hide_default'] = 0;

				foreach (array('sync', 'post_types', 'taxonomies') as $key)
					$this->options[$key] = empty($_POST[$key]) ? array() : array_keys($_POST[$key], 1);

				update_option('easyMultilingual', $this->options);

				// refresh rewrite rules in case rewrite,  hide_default, post types or taxonomies options have been modified
				// it seems useless to refresh permastruct here
				flush_rewrite_rules();

				// refresh language cache in case home urls have been modified
				$this->model->clean_languages_cache();

				// fills existing posts & terms with default language
				if (isset($_POST['fill_languages']) && $nolang = $this->model->get_objects_with_no_lang()) {
					if (!empty($nolang['posts']))
						$this->model->set_language_in_mass('post', $nolang['posts'], $this->options['default_lang']);
					if (!empty($nolang['terms']))
						$this->model->set_language_in_mass('term', $nolang['terms'], $this->options['default_lang']);
				}

				add_settings_error('general', 'settings_updated', __('Settings saved.'), 'updated');
				$this->redirect();
				break;

			default:
				break;
		}

		// displays the page
		include(EML_ADMIN_INC.'/view-languages.php');
	}

	/*
	 * register strings for translation making sure it is not duplicate or empty
	 *
	 * @since 0.6
	 *
	 * @param string $name a unique name for the string
	 * @param string $string the string to register
	 * @param string $context optional the group in which the string is registered, defaults to 'easyMultilingual'
	 * @param bool $multiline optional wether the string table should display a multiline textarea or a single line input, defaults to single line
	 */
	public function register_string($name, $string, $context = 'easyMultilingual', $multiline = false) {
		// backward compatibility with EasyMultilingual older than 1.1
		if (is_bool($context)) {
			$multiline = $context;
			$context = 'easyMultilingual';
		}

		$to_register = compact('name', 'string', 'context', 'multiline');
		if (!in_array($to_register, $this->strings) && $to_register['string'])
			$this->strings[] = $to_register;
	}

	/*
	 * get registered strings
	 *
	 * @since 0.6.1
	 *
	 * @return array list of all registered strings
	 */
	public function &get_strings() {
		// WP strings
		$this->register_string(__('Site Title'), get_option('blogname'), 'WordPress');
		$this->register_string(__('Tagline'), get_option('blogdescription'), 'WordPress');
		$this->register_string(__('Date Format'), get_option('date_format'), 'WordPress');
		$this->register_string(__('Time Format'), get_option('time_format'), 'WordPress');

		// widgets titles
		global $wp_registered_widgets;
		$sidebars = wp_get_sidebars_widgets();
		foreach ($sidebars as $sidebar => $widgets) {
			if ($sidebar == 'wp_inactive_widgets' || empty($widgets))
				continue;

			foreach ($widgets as $widget) {
				// nothing can be done if the widget is created using pre WP2.8 API :(
				// there is no object, so we can't access it to get the widget options
				if (!isset($wp_registered_widgets[$widget]['callback'][0]) || !is_object($wp_registered_widgets[$widget]['callback'][0]) || !method_exists($wp_registered_widgets[$widget]['callback'][0], 'get_settings'))
					continue;

				$widget_settings = $wp_registered_widgets[$widget]['callback'][0]->get_settings();
				$number = $wp_registered_widgets[$widget]['params'][0]['number'];
				// don't enable widget title translation if the widget is visible in only one language or if there is no title
				if (empty($widget_settings[$number]['eml_lang']) && isset($widget_settings[$number]['title']) && $title = $widget_settings[$number]['title'])
					$this->register_string(__('Widget title', 'easyMultilingual'), $title, 'Widget');
			}
		}

		// allow plugins to modify our list of strings, mainly for use by our EML_WPML_Compat class
		$this->strings = apply_filters('eml_get_strings', $this->strings);
		return $this->strings;
	}

	/*
	 * list the post metas to synchronize
	 *
	 * @since 1.0
	 *
	 * @return array
	 */
	static public function list_metas_to_sync() {
		return array(
			'taxonomies'        => __('Taxonomies', 'easyMultilingual'),
			'post_meta'         => __('Custom fields', 'easyMultilingual'),
			'comment_status'    => __('Comment status', 'easyMultilingual'),
			'ping_status'       => __('Ping status', 'easyMultilingual'),
			'sticky_posts'      => __('Sticky posts', 'easyMultilingual'),
			'post_date'         => __('Published date', 'easyMultilingual'),
			'post_format'       => __('Post format', 'easyMultilingual'),
			'post_parent'       => __('Page parent', 'easyMultilingual'),
			'_wp_page_template' => __('Page template', 'easyMultilingual'),
			'menu_order'        => __('Page order', 'easyMultilingual'),
			'_thumbnail_id'     => __('Featured image', 'easyMultilingual'),
		);
	}

	/*
	 * redirects to language page (current active tab)
	 * saves error messages in a transient for reuse in redirected page
	 *
	 * @since 1.5
	 *
	 * @param array $args query arguments to add to the url
	 */
	protected function redirect($args = array()) {
		if ($errors = get_settings_errors()) {
			set_transient('settings_errors', $errors, 30);
			$args['settings-updated'] = 1;
		}
		wp_redirect(add_query_arg($args,  wp_get_referer() ));
		exit;
	}
}
